### PR TITLE
Add Multiraft transport implementation.

### PR DESCRIPTION
**Pull request by tnachen**
I haven't added any test, but this is my first patch for cockroach (and pretty much my foray into golang).
So want to put up a early review to make sure I'm not going down a wrong path.

### CODE REVIEW

**Comment by tnachen** [comment]
@bdarnell I wasn't sure how does the multiraft server id translates into RaftMessageRequest, I assume it is the sender/to info.
And also what's your recommended way to test this?

File: multiraft/transport.go [hunk 1]
```diff
@@ -23,6 +23,8 @@ import (
 	"strings"
 	"sync"
 
+	"github.com/cockroachdb/cockroach/gossip"
```
Comment (bdarnell) [inline multiraft/transport.go]: multiraft shouldn't import gossip (it is at least in principle independent from the rest of cockroach). This cockroach-specific RPCTransport should live elsewhere, probably either cockroach/rpc or cockroach/server.

**Comment by bdarnell** [comment]
The change looks good aside from where the new code will live. For testing, I think we need something that starts up two rpc servers (hopefully this can be done with just a net/rpc server and a dummy channel-based implementation of ServerInterface, instead of pulling in the rest of cockroach) and two RPCTransports, gossips their addresses, and sends some messages back and forth.

We'll also want to wire this in to the storage/raft.go interfaces (although I wouldn't try to do that in the same PR). If we can get rid of singleNodeRaft completely and just use this that would be ideal, although I don't know if we have enough of the gossip machinery running in all the tests to do that. This will probably inform what package the code will live in - we want to be able to import RPCTransport from storage (I think? or maybe we just pass it in when we construct the Store), which rules out cockroach/server.

**Comment by bdarnell** [comment]
FYI I'm going to start working on the storage/raft.go changes I referred to above - replace singleNodeRaft with something that can address multiple nodes, although still using LocalRPCTransport for now.

**Comment by tnachen** [comment]
Hi Ben,
Sounds good, sorry I am on vacation and traveling overseas until 2/7 so couldn't get
As much as I hope for in this PR.
My latest commit is still pending test,
Let me know how I can help

**Comment by bdarnell** [comment]
OK. This isn't blocking anything yet so we can wait until you're back to finish it up.

**Comment by tnachen** [comment]
@bdarnell Can you take a look? I've added one test and moved to another package.

File: server/transport_test.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+	transport2, err := NewRPCTransport(gossip, rpcServer2)
+	if err != nil {
+		t.Errorf("Unexpected error creating transport2, Error: %s", err)
+	}
+
+	gossip.AddInfo("node-1", rpcServer1.Addr(), time.Hour)
+	gossip.AddInfo("node-2", rpcServer2.Addr(), time.Hour)
+
+	testServer := &TestTransportServer{}
+	transport1.Listen(1, testServer)
```
Comment (bdarnell) [inline server/transport_test.go]: How does this work? You're only binding one TestTransportServer but both nodes are able to connect. I don't think Connect is a strong enough test; you should actually send messages between the two servers (add a channel to TestTransportServer to get the messages out).

Comment (tnachen) [inline server/transport_test.go]: Good point, I was going to create another test but let me modify this one. I'm still figuring out how to get the channel to work.

**Comment by bdarnell** [comment]
Replaced by #483

**Pull request closed without merging**